### PR TITLE
33267

### PR DESCRIPTION
**Pull request by TrucLuu**
Admin feature

### CODE REVIEW

File: app/controllers/admin/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
+class Admin::UsersController < ApplicationController
+  before_action :set_user, only: [:show, :edit, :update, :destroy, :active, :admin]
```
Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: not use only. 
:active what ?

File: app/controllers/admin/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    @user = User.new
+  end
+
+  def edit
+  end
+
+  def create
+    @user = User.new admin_user_params
+    if @user.save
+      redirect_to [:admin, @user], notice: t("admin.user.update.success")
```
Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: Remove ( )

File: app/controllers/admin/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    else
+      render :new
+    end
+  end
+
+  def update
+    if @user.update admin_user_params
```
Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: update => update_attributes

Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: admin_user_params => user_params

File: app/controllers/admin/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    redirect_to admin_users_url
+  end
+
+  private
+  def set_user
+    @user = User.find params[:id]
+  end
+
+  def admin_user_params
+    params.require(:user).permit(
```
Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: Remove ( )

Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: Use once line

File: app/views/admin/users/edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<h1>
+  <%= "#{t("features.edit")}
+      #{t("title.admin_title")}
```
Comment (phamkykhoi) [inline app/views/admin/users/edit.html.erb]: Indent

File: app/views/admin/users/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+      #{t("login.user")}" %>
```
Comment (phamkykhoi) [inline app/views/admin/users/edit.html.erb]: Indent

File: app/views/admin/users/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  <tbody>
+    <% @users.each do |user| %>
+      <tr>
+        <td><%= user.name %></td>
+        <td><%= user.email %></td>
+        <td><%= user.active %></td>
+        <td><%= user.admin %></td>
+        <td><%= link_to t("features.show") %></td>
+        <td><%= link_to t("features.edit"), edit_admin_user_path(user) %></td>
+        <td><%= link_to t("features.destroy"), user, method: :delete, data: { confirm: t("login.question") } %></td>
```
Comment (phamkykhoi) [inline app/views/admin/users/index.html.erb]: Remove space

File: app/views/layouts/_header.html.erb [hunk 8]
```diff
@@ -19,6 +19,21 @@
               </li>
             </ul>
           </li>
+          <% if current_user.is_admin? %>
+            <li class="dropdown">
+              <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+                Admin <b class="caret"></b>
```
Comment (phamkykhoi) [inline app/views/layouts/_header.html.erb]: I18n

File: app/controllers/admin/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    @user.destroy
+    redirect_to admin_users_url
+  end
+
+  private
+  def set_user
+    @user = User.find params[:id]
+  end
+
+  def admin_user_params
```
Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: Change name to user_params

File: app/controllers/admin/users_controller.rb (hunk 1 above)
Comment (TrucLuu) [inline app/controllers/admin/users_controller.rb]: `active` use for active account admin

File: app/controllers/admin/users_controller.rb (hunk 4 above)
Comment (TrucLuu) [inline app/controllers/admin/users_controller.rb]: More 80 character..

File: app/views/admin/users/index.html.erb (hunk 7 above)
Comment (TrucLuu) [inline app/views/admin/users/index.html.erb]: Can't not remove `{}` My app will not running.

File: app/controllers/admin/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  def create
+    @user = User.new user_params
+    if @user.save
+      redirect_to [:admin, @user], notice: t("admin.user.update.success")
+    else
+      render :new
+    end
+  end
+
+  def update_attributes
```
Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: change name function to update

File: app/controllers/admin/users_controller.rb [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    if @user.update user_params
```
Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: @user.update user_params => @user.update_attributes user_params

Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: Insert line above @user = User.find params[:id]

Comment (TrucLuu) [inline app/controllers/admin/users_controller.rb]: https://github.com/TrucLuu/fels_63/blob/33267/app/controllers/admin/users_controller.rb#L2

File: app/views/admin/users/edit.html.erb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<h1>
+  <%= "#{t("features.edit")} #{t("title.admin_title")} #{t("login.user")}" %>
+</h1>
+
+<%= render 'form' %>
```
Comment (nguyenngoctruong) [inline app/views/admin/users/edit.html.erb]: ' => "

File: app/controllers/admin/users_controller.rb [hunk 13]
```diff
@@ -0,0 +1,56 @@
+class Admin::UsersController < ApplicationController
+  before_action :set_user, only: [:show, :edit, :update, :destroy]
```
Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: Change only to except: [:index, :new, :create]

File: app/views/admin/users/index.html.erb [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    <% @users.each do |user| %>
+      <tr>
+        <td><%= user.name %></td>
+        <td><%= user.email %></td>
+        <td><%= user.active %></td>
+        <td><%= user.admin %></td>
+        <td><%= link_to t("features.show") %></td>
+        <td><%= link_to t("features.edit"), edit_admin_user_path(user) %></td>
+        <td><%= link_to t("features.destroy"), user, method: :delete,
+                        data: { confirm: t("login.question") } %></td>
```
Comment (nguyenngoctruong) [inline app/views/admin/users/index.html.erb]: bỏ dấu cách 2 đầu ngoặc ngọn anh ơi

File: app/controllers/admin/users_controller.rb [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    else
+      render :new
+    end
+  end
+
+  def update
+    if @user.update_attributes user_params
+      redirect_to [:admin, @user], notice: t("admin.user.update.success")
+    else
+      render :edit
```
Comment (phamkykhoi) [inline app/controllers/admin/users_controller.rb]: :edit => "edit"

File: app/controllers/admin/users_controller.rb (hunk 13 above)
Comment (TrucLuu) [inline app/controllers/admin/users_controller.rb]: I'm using rails 4.2.1. At there version does not use `except`.

File: app/views/admin/users/new.html.erb [hunk 16]
```diff
@@ -0,0 +1,5 @@
+<h1><%= t("admin.user.new") %></h1>
+
+<%= render 'form' %>
```
Comment (nguyenngoctruong) [inline app/views/admin/users/new.html.erb]: ' => "

File: app/views/admin/users/_form.html.erb [hunk 17]
```diff
@@ -0,0 +1,41 @@
+<%= form_for [:admin, @user] do |f| %>
+  <% if @user.errors.any? %>
```
Comment (phamkykhoi) [inline app/views/admin/users/_form.html.erb]: Use file share/_error.html.erb

File: app/controllers/admin/users_controller.rb [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+  def update
+    if @user.update_attributes user_params
+      redirect_to [:admin, @user], notice: t("admin.user.update.success")
+    else
+      render "edit"
+    end
+  end
+
+  def destroy
+    @user.destroy
```
Comment (khanhkhanh) [inline app/controllers/admin/users_controller.rb]: If .. else .. end

File: app/helpers/sessions_helper.rb [hunk 19]
```diff
@@ -58,3 +58,10 @@ def forget user
     cookies.delete :remember_token
   end
 end
+
+def admin_auth
+  logged_in_user
+  unless current_user.is_admin?
+    redirect_to root_url, danger: t("admin.not_authenticated")
+  end
+end
```
Comment (khanhkhanh) [inline app/helpers/sessions_helper.rb]: Break line

File: app/views/admin/users/_form.html.erb [hunk 20]
```diff
@@ -0,0 +1,31 @@
+<%= form_for [:admin, @user] do |f| %>
+  <%= render 'shared/error_messages', object: f.object %>
```
Comment (khanhkhanh) [inline app/views/admin/users/_form.html.erb]: ' => "

File: app/models/user.rb [hunk 21]
```diff
@@ -87,4 +86,8 @@ def send_password_reset_email
   def password_reset_expired?
     reset_sent_at < Settings.TWOHOUR.hours.ago
   end
+
+  def is_admin?
+    self.active && self.admin
```
Comment (khanhkhanh) [inline app/models/user.rb]: self.active? && self.admin?

File: app/views/admin/users/index.html.erb [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+      <th><%= t("title.name") %></th>
+      <th><%= t("title.email") %></th>
+      <th><%= t("title.active") %></th>
+      <th><%= t("title.admin_title") %></th>
+      <th colspan="3"></th>
+    </tr>
+  </thead>
+
+  <tbody>
+    <% @users.each do |user| %>
```
Comment (khanhkhanh) [inline app/views/admin/users/index.html.erb]: Use render instead

File: app/views/admin/users/index.html.erb [hunk 23]
```diff
@@ -0,0 +1,33 @@
+<p id="notice"><%= notice %></p>
+
+<h1><%= t("admin.user.list") %></h1>
```
Comment (khanhkhanh) [inline app/views/admin/users/index.html.erb]: Remove ()

File: app/views/admin/users/index.html.erb [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+
+<table class="table table-bordered table-striped">
+  <thead>
+    <tr>
+      <th><%= t("title.name") %></th>
```
Comment (khanhkhanh) [inline app/views/admin/users/index.html.erb]: Remove ()

File: app/views/admin/users/new.html.erb [hunk 25]
```diff
@@ -0,0 +1,5 @@
+<h1><%= t("admin.user.new") %></h1>
```
Comment (khanhkhanh) [inline app/views/admin/users/new.html.erb]: Remove ()

File: app/views/admin/users/index.html.erb (hunk 24 above)
Comment (khanhkhanh) [inline app/views/admin/users/index.html.erb]: Remove () if possible

File: app/views/admin/users/index.html.erb (hunk 23 above)
Comment (khanhkhanh) [inline app/views/admin/users/index.html.erb]: Remove () if possible

File: app/views/admin/users/new.html.erb (hunk 25 above)
Comment (khanhkhanh) [inline app/views/admin/users/new.html.erb]: Remove () if possible

File: app/views/admin/users/show.html.erb [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+  <%= @user.active %>
+</p>
+
+<p>
+  <strong><%= t "title.admin_title" %>:</strong>
+  <%= @user.admin %>
+</p>
+
+<%= link_to  t("features.edit") , edit_admin_user_path(@user) %> |
+<%= link_to t("features.back"), admin_users_path %>
```
Comment (khanhkhanh) [inline app/views/admin/users/show.html.erb]: Break line

File: app/views/admin/users/show.html.erb [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+  <strong><%= t "title.active" %>:</strong>
+  <%= @user.active %>
+</p>
+
+<p>
+  <strong><%= t "title.admin_title" %>:</strong>
+  <%= @user.admin %>
+</p>
+
+<%= link_to  t("features.edit") , edit_admin_user_path(@user) %> |
```
Comment (khanhkhanh) [inline app/views/admin/users/show.html.erb]: Remove spaces before and after t("features.edit")

File: app/views/layouts/_header.html.erb [hunk 28]
```diff
@@ -9,7 +9,7 @@
           <li><%= link_to t("login.user"), users_path %></li>
           <li class="dropdown">
             <a href="#" class="dropdown-toggle" data-toggle="dropdown">
-              <%= t("login.account") %> <b class="caret"></b>
+              <%= t("login.account") %><b class="caret"></b>
```
Comment (khanhkhanh) [inline app/views/layouts/_header.html.erb]: Remove () if possible

File: app/views/layouts/_header.html.erb [hunk 29]
```diff
@@ -19,6 +19,21 @@
               </li>
             </ul>
           </li>
+          <% if current_user.is_admin? %>
+            <li class="dropdown">
+              <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+                <%= t("title.admin_title") %> <b class="caret"></b>
```
Comment (khanhkhanh) [inline app/views/layouts/_header.html.erb]: Remove () if possible

File: app/controllers/admin/users_controller.rb (hunk 18 above)
Comment (TrucLuu) [inline app/controllers/admin/users_controller.rb]: I don't understand your mind.. Pls, explain to me. Thank

**Pull request merged**